### PR TITLE
DRILL-8136: Overload existing Math UDFs to allow for VARCHAR input

### PR DESCRIPTION
**Pull request by estherbuchwalter**
# [DRILL-8136](https://issues.apache.org/jira/browse/DRILL-8136): Overload existing Math UDFs to allow for VARCHAR input

## Description

Wrote UDFs to overload existing Math functions in Drill so that Math UDFs accept input in type VARCHAR.

## Documentation
All functions [here](https://drill.apache.org/docs/math-and-trig/) that accept input, except for the bitwise functions, are overloaded.

## Testing
Tested each function with unit tests in `TestMathFunctionsVarcharInput.java`

### CODE REVIEW

**Review by cgivre — COMMENTED** [review]
@estherbuchwalter 
This looks good.  One question was why did you use `VarCharHolders` instead of `NullableVarCharHolder`?   For the purpose of making this durable, it might be best to replace the `VarCharHolders` with the nullable versions so that null values don't cause crashes.   Maybe write a unit test or two with `null` values now and see what works best.

**Comment by estherbuchwalter** [comment]
Ok, thank you for the review. I used `VarCharHolders` because that's what had been used for the Math UDFs that I am overloading.
Yes, I'll add some unit tests with `null` values. Also, I'm going to include a check for `null` in the functions so that `null` values don't cause crashes.

**Comment by cgivre** [comment]
@estherbuchwalter Please don't forget the license headers so that this PR passes the CI.

**Comment by estherbuchwalter** [comment]
@cgivre Thanks for the reminder about the headers!

**Comment by estherbuchwalter** [comment]
As of now the random() method that accepts a seed only accepts input of type `int`. Does it make sense to allow the random(int seed) to accept type `double`/`decimal` as well?

**Comment by paul-rogers** [comment]
Thanks for the contribution. Please explain the casting rules assumed here. Is the VARCHAR assumed to be an INT? BIGINT? DOUBLE? DECIMAL? That assumption affects type propagation, and is why normal SQL requires the user to specify the type via a CAST operation.

**Comment by estherbuchwalter** [comment]
The functions are designed to accept a VARCHAR which can be an INT, BIGINT, DOUBLE, or DECIMAL.

**Comment by paul-rogers** [comment]
@estherbuchwalter, thanks for the explanation. The function must, however, declare an output type. The other input must be cast to the correct type. For example, if the VARCHAR is to be cast to a DECIMAL, so must the other argument, and the output is DECIMAL.

On the other hand, if we infer the type from the numeric type, then 10 + "10.1" will produce 20, not 20.1.

SQL is a statically typed language: every row must produce the output type. Drill does defer the type decision to runtime, but the type decision becomes frozen on the first row.

To be clear, what is the output type for:

VARCHAR + INT --> ?
VARCHAR + DOUBLE --> ?
VARCHAR + DECIMAL -> ?

Since the VARCHAR can be anything, the safest result is if the output is always VARDECIMAL (unlimited precision) or DOUBLE (largest range of values, with some truncation). For any other output type, we can invent an input that will make the output type invalid.

Again, this is why SQL normally requires:

(CAST col AS INT) + 10 --> BIGINT

So that the user declares that the VARCHAR "col" column is converable to INT, and thus the normal INT addition type promotion rules apply.

Or, have you come up with some trick to work around this issue?

**Comment by estherbuchwalter** [comment]
@paul-rogers, thank you for your question. Please let me know if this explanation satisfies your question. 

In order to maintain precision, each input is parsed to Java type `DOUBLE` (with a few exceptions). I am working on minimizing, or if possible, completely getting rid of the exceptions.
First there is a check to ensure that the input is a valid number (accepting `INT` or `DOUBLE`), and then the input is parsed to a `DOUBLE`.
Then, the Math function is performed and the function's output is a `Float8Holder`, except for the `sign()` function where the output is an `IntHolder`.
Are you suggesting that it would be better to return a `VARDECIMAL` than a `Float8Holder`?

**Comment by estherbuchwalter** [comment]
To clarify, the purpose of this PR is to make Drill more resilient with different types of input to Math functions. The goal here is that Drill should not throw an error, and thereby put a halt on all of the query results, when one value within a large dataset is a different data type than what the Math function originally expected. Instead it should either know how to handle the different data type or return a value that conveys that the input was invalid (like `null`, `NaN`, etc).
For example, before this PR the following SQL query would send a Drill error: `SELECT abs('-10')`. However, with this PR, the correct value would be returned (it would be returned as a Float8Holder).

Also, it seems that this functionality would be consistent with industry standards. I have tried different types of input to Math functions in MySQL and Postgres databases and each of them accepts VARCHAR input (although they do handle it a little differently from each other).

File: exec/java-exec/src/test/java/org/apache/drill/exec/expr/fn/impl/TestMathFunctionsVarcharInput.java [hunk 1]
```diff
@@ -88,7 +88,7 @@ public void testModVarcharInput() throws Exception {
       "FROM (values (1))";
 
     RowSet results = client.queryBuilder().sql(sql).rowSet();
-    
+
```
Comment (estherbuchwalter) [inline exec/java-exec/src/test/java/org/apache/drill/exec/expr/fn/impl/TestMathFunctionsVarcharInput.java:91]: I don't see any change here. Am I missing something?

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctionsVarcharUtils.java [hunk 2]
```diff
@@ -18,15 +18,24 @@
...
+  public static boolean isValid(String input) {
+    if (StringUtils.isEmpty(input)) {
+      return false;
+    } else {
+      input = input.trim();
+      return NumberUtils.isCreatable(input);
+    }
+  }
 
+  public static String validateInput(String input) {
```
Comment (estherbuchwalter) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctionsVarcharUtils.java:38]: @cgivre, thanks for adding the `NumberUtils.isCreatable` function to check for a valid number. Should I get rid of the `validateInput()` method?

Comment (cgivre) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctionsVarcharUtils.java:38]: Yeah, I don't think it's necessary.

**Comment by cgivre** [comment]
I had a thought... one thing we might want to try is to change the null handling from `NULL_IF_NULL` to `INTERNAL` and then see if we can get the functions to return `null` in the event of unparsable input.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctions.java [hunk 3]
```diff
@@ -33,22 +33,40 @@
...
+    @Param VarCharHolder b;
     @Output  Float8Holder out;
 
-    public void setup(){}
+    @Workspace org.apache.drill.exec.expr.fn.impl.MathFunctionsVarcharUtils mathUtils;
 
-    public void eval(){
-      out.value = java.lang.Math.pow(a.value, b.value);
+    public void setup(){
+      mathUtils = new MathFunctionsVarcharUtils();
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctions.java:47]: Would recommend using static functions rather an instance and member variables.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctions.java [hunk 4]
```diff
@@ -33,22 +33,40 @@
 import org.apache.drill.exec.expr.holders.VarCharHolder;
 
 public class MathFunctions {
-
+/*
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctions.java:36]: Why is this commented out and why does it replace the existing Float8 implementation?

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctionsVarcharInput.java [hunk 5]
```diff
@@ -0,0 +1,569 @@
...
+import org.apache.drill.exec.expr.annotations.Param;
+import org.apache.drill.exec.expr.annotations.Workspace;
+import org.apache.drill.exec.expr.holders.Float8Holder;
+import org.apache.drill.exec.expr.holders.IntHolder;
+import org.apache.drill.exec.expr.holders.NullableFloat8Holder;
+import org.apache.drill.exec.expr.holders.NullableVarCharHolder;
+import org.apache.drill.exec.expr.holders.VarCharHolder;
+
+public class MathFunctionsVarcharInput {
+  /*
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctionsVarcharInput.java:34]: Why are we adding new code which is commented out?

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctionsVarcharInput.java [hunk 6]
```diff
@@ -0,0 +1,569 @@
...
+    }
+
+   */
+  @FunctionTemplate(name = "mod",
+    scope = FunctionTemplate.FunctionScope.SIMPLE,
+    nulls = FunctionTemplate.NullHandling.NULL_IF_NULL)
+  public static class Mod implements DrillSimpleFunc {
+
+    @Param
+    VarCharHolder a;
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctionsVarcharInput.java:135]: These functions are a bit tricky: I wonder if there is a bit of misunderstanding. Drill uses a typed system at execution time. Each math function has declared input types: BIGINT, DOUBLE, VARDECIMAL, etc. The type promotion system will pick the "narrowest" type that is compatible with both arguments, then will cast both arguments, as needed to widen them to the target type. So MOD(BIGINT, BIGINT) is compatible with MOD(B, I) where B is a BIGINT, and I is INT. I is then cast to BIGINT.

As a result, we generate MANY overrides of the same function: one for each of the numeric "families": BIGINT, DOUBLE, VARDECIMAL, with sometimes others thrown in as well, such as maybe an INT version.

So, let's consider this improvement. We want a MOD(VARCHAR, X). What is X? In this bit of code, X is VARCHAR, which matches MOD("10.3", "20"). Cool. But, what about MOD("10.3", 20)? I don't think Drill automatically promotes from numeric types to VARCHAR. (Didn't test, but few SQL systems do.) So, you'd also want MOD(VARCHAR, VARDECIMAL), so the second argument can be cast from the integer and floating point types. To heavy-weight, then add MOD(VARCHAR, BIGINT) and MOD(VARCHAR, DOUBLE). Now the bases are covered.

Except, not yet. The above only handles the second argument. What about the first? So, we also need MOD(BIGINT, VARCHAR), MOD(DOUBLE, VARCHAR), MOD(VARDECIMAL, VARCHAR).

That's a total of 7 implementations for each two-argument math function!

If we did that, I'd almost suggest we break SQL type casting rules and define automatic conversions from VARCHAR to any numeric type. We pick the type based on the auto-cast of VARCHAR (to DOUBLE, say) and any numeric argument. In this case, you get to reuse all the existing math functions for free: no need to change the casting system the hard way.

The result would no longer be SQL compatible. So, maybe a system/session option to enable "Python mode" when needed.

Comment (estherbuchwalter) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctionsVarcharInput.java:135]: @paul-rogers, thank you for your input. I appreciate your comments and thorough explanation. 
We are thinking about these issues and experimenting to see what would work well with Drill and accomplish our goal. 

Please note that this PR is currently marked as `DRAFT` since it is very much a WIP. We can request a review when it is ready.

**Comment by cgivre** [comment]
This PR is not necessary once https://github.com/apache/drill/pull/2638 is merged.
